diff --git a/src/afs/LINUX/osi_file.c b/src/afs/LINUX/osi_file.c
index c125c1c..e211c05 100644
--- a/src/afs/LINUX/osi_file.c
+++ b/src/afs/LINUX/osi_file.c
@@ -183,7 +183,11 @@ osi_UFSTruncate(struct osi_file *afile, afs_int32 asize)
     newattrs.ia_ctime = CURRENT_TIME;
 
     /* avoid notify_change() since it wants to update dentry->d_parent */
+#ifdef HAVE_LINUX_SETATTR_PREPARE
+    code = setattr_prepare(afile->filp->f_path.dentry, &newattrs);
+#else
     code = inode_change_ok(inode, &newattrs);
+#endif
     if (!code)
        code = afs_inode_setattr(afile, &newattrs);
     if (!code)

