--- open-vm-tools/configure.ac.orig     2015-10-08 09:46:41.945500113 +0200
+++ open-vm-tools/configure.ac  2015-10-08 09:48:51.312502451 +0200
@@ -1049,8 +1049,8 @@
    if test "$osVersion" -lt 303000; then
       MODULES="$MODULES vmxnet"
    fi
-   # See if we need vmhgfs module. Starting with 4.0.0 we use FUSE
-   if test "$osVersion" -lt 400000; then
+   # See if we need vmhgfs module. Starting with 4.2.0 we use FUSE
+   if test "$osVersion" -lt 402000; then
       MODULES="$MODULES vmhgfs"
    fi
    # See if we need vmci and vsock modules. Starting with 3.9 they made
